### PR TITLE
VM publish & unified images import

### PR DESCRIPTION
**Pull request by stgraber**

### CODE REVIEW

**Comment by stgraber** [comment]
Unless I missed something, all that's left to do is hook up a tar reader to directly extract the qcow2 and feed it to qemu-img. Hopefully this can be done somewhat easily with Go's tar implementation.

A further optimization there would be to actually split the entire unpack process so for VMs we use the Go tar reader for everything (only need to deal with compression once) but keep using the out of process tar for containers where we definitely prefer it when unpacking a rootfs.

**Comment by stgraber** [comment]
Sorted out the importing code path, all working now.

File: lxd/images.go [hunk 1]
```diff
@@ -173,7 +173,7 @@ func imgPostContInfo(d *Daemon, r *http.Request, req api.ImagesPost, op *operati
 		if !shared.IsSnapshot(name) {
 			return nil, fmt.Errorf("Not a snapshot")
 		}
-	case "container":
+	case "container", "virtual-machine", "instance":
```
Comment (tomponline) [inline lxd/images.go]: @stgraber should we validate the ctype specified with the instance type loaded on line 192?

Comment (stgraber) [reply lxd/images.go]: No as our current client isn't very smart and always passes container :)

Comment (stgraber) [reply lxd/images.go]: I think I'll send a follow-up fix to have the client always pass "instance" so long as we have the instances API extension, will make things a bit less wrong there :)

Comment (stgraber) [reply lxd/images.go]: (for speed reasons, the client doesn't even pull the instance so it doesn't know what the type is)

Comment (tomponline) [reply lxd/images.go]: @stgraber ok makes sense, and that sounds good. Thanks

**Pull request merged**